dspMetadataGUI/util/metaDataSet.py

```diff
@@ -0,0 +1,1290 @@
+        """
+        try:
+            ttl = self.turtle
+            if not isinstance(ttl, str):
+                raise Exception('turtle was not a string')
+            if ttl:
+                return ttl
+            else:
+                raise Exception('no turtle found')
+        except Exception as e:
+            print('Warning: No turtle cached. Performance may be decreased.')
+            print(f'    (Exception: {e})')
+            # LATER: remove with next breaking change
+            g = self.generate_rdf_graph()
+            return g.serialize(format='turtle').decode("utf-8")
+
+
+class DataClass(ABC):
+    """
+    Abstract parent class of all classes holding data.
+    """
+
+    def get_metadataset(self) -> MetaDataSet:
+        """
+        Returns the `MetaDataSet` to which this class belongs
+
+        Returns:
+            MetaDataSet: The owner MetaDataSet
+        """
+        return self.meta
+
+    def add_rdf_to_graph(self, graph: Graph, typename: str):
+        """
+        Adds the RDF representation of this class to a graph
+
+        Args:
+            graph (Graph): the `rdflib.Graph` to which the triples should be added
+            typename (str): the type name of this class
+        """
+        iri = self.get_rdf_iri()
+        type = dsp_repo[typename]
+        graph.add((iri, RDF.type, type))
+        for prop in self.get_properties():
+            graph += prop.get_triples(iri)
+
+    def get_rdf_iri(self) -> URIRef:
+        """
+        Return the IRI of the object.
+
+        This can be called to use the object as subject of an RDF triple.
+
+        Returns:
+            URIRef: the IRI
+        """
+        shortcode = self.get_metadataset().project.shortcode.value
+        if not shortcode:
+            shortcode = "xxxx"
+        classname = 'dsp-' + shortcode + self.iri_suffix
+        return URIRef(dsp_repo[classname])
+
+    @abstractmethod
+    def get_properties(self) -> list:
+        """
+        Return a list of all `Property` fields of this class
+        """
+        raise NotImplementedError
+
+    def get_prop_by_name(self, name):
+        for p in self.get_properties():
+            if p.name == name:
+                return p
+
+
+class Project(DataClass):
+    """
+    Project shape.
+
+    Corresponds to `dsp-repo:Project` in our ontology.
+
+    Args:
+        name (str): The name of the project
+        shortcode (str): The project shortcode
+        meta (MetaDataSet): the owning `MetaDataSet`
+    """
+
+    def __init__(self, name: str, shortcode: str, meta: MetaDataSet):
+        self.meta = meta
+        self.iri_suffix = "-project"
+        self.name = Property(meta, "Name",
+                             "The name of the Project",
+                             "Test Project",
+                             Datatype.STRING,
+                             Cardinality.ONE,
+                             name,
+                             predicate=dsp_repo.hasName)
```

Comment:
`self.name` is of type str but value of type "Property" is assigned

Reply:
no, `self.name` is supposed to be of type `Property`.

I'm afraid the variable and property names make it confusing, here:  
Class `Project` has a Property `name` of type `Property`; but at the same time, class `Property` has a property of type `str`.  
So at the end of the day, `Project` has a property `name` which has a the name "name".  
It makes more sense when you think of other cases: `Project` also has a property `description` with the name "description".

Does that make sense now?

Reply:
You're right. I didn't see that